src/bijectors/ordered.jl

```diff
@@ -17,14 +25,22 @@ Return a `Distribution` whose support are ordered vectors, i.e., vectors with in
 This transformation is currently only supported for otherwise unconstrained distributions.
 """
 function ordered(d::ContinuousMultivariateDistribution)
-    if bijector(d) !== identity
-        throw(
-            ArgumentError(
-                "ordered transform is currently only supported for unconstrained distributions.",
-            ),
-        )
+    # We're good if the map from unconstrained (in which we apply the ordered bijector)
```

Comment:
Would it be useful to add to the docstring a comment about when this is safe to use? e.g. in general, if `d` has variable parameters, then this is unsafe, because it implicitly renormalizes the distribution, but the log of the new normalization constant is never computed so is never included in the log density or gradient, though it would need to be when the parameters are variable.

Reply:
When you say "variable parameters", do you mean that the size / length can vary?

Reply:
No, I mean parameters that are being sampled, optimized in the model.

Reply:
So you're saying
```julia
sigma ~ InverseGamma(2, 3)
x ~ ordered(MvNormal(zeros(2), sigma^2 * I))
```
will be incorrect?

Reply:
> because it implicitly renormalizes the distribution, but the log of the new normalization constant is never computed so is never included in the log density or gradient,

Okay so I understand what you mean by this now, but I'm struggling a bit to understand why this is an issue here but not for other "typical" diffeomorphisms.

Reply:
This technically would be an issue for other diffeomorphisms. E.g. if we used an Exp bijector for a normal random variable. The difference is that we generally define a bijector for a distribution already, so we reduce the chance this happens. E.g. in this case the user would construct truncated(Normal(m,s); lower=0), and then the normalization factor is computed. With ordered, the user is choosing to apply an additional ordered bijector, which then restricts the support.

FWIW, the docstring for ordered doesn't clarify whether it constructs the restriction of the wrapped distribution to the subset of the support consisting of ordered vectors (i.e. sample with rejection sampling) or whether it constructs the distribution whose random sampler consists of sampling from the wrapped distribution and then sorting. It should be the first.

Reply:
> This technically would be an issue for other diffeomorphisms. E.g. if we used an Exp bijector for a normal random variable.

I still don't understand this part. The jacobian factor should make it so that everything is still normalized, no? So we're taking a normalized distribution and re-weighting it in a way to preserve normalization under the transformation.

> FWIW, the docstring for ordered doesn't clarify whether it constructs the restriction of the wrapped distribution to the subset of the support consisting of ordered vectors (i.e. sample with rejection sampling) or whether it constructs the distribution whose random sampler consists of sampling from the wrapped distribution and then sorting. It should be the first.

This I very much agree with :+1:

Reply:
Well another way to put it is that the function is not in fact a diffeomorphisms, because it's injective, not bijective. The original distribution is defined on some support, but `ordered(...)` has a support that is a subset of the original distribution. So we're not just applying a bijective map and worrying about the change of measure. We're additionally restricting the support, which the Jacobian cannot capture. If we had a magic function that could compute normalization factors, then we would use that to compute the normalized logpdf after restricting the support, and then Ordered would indeed be a bijection.

Reply:
Same deal. Because you're taking a density whose support is the whole real line and using a change of variables from the real line to the positive reals (not the whole real line). So you're changing the support if you use `Exp` for this distribution.

Reply:
Nah, then I still don't get it. I'm with you that you're changing the domain, but $\exp: \mathbb{R} \to (0, \infty)$ is bijective, which are the domains we are interested in. Why does it matter that $\exp$ treated as a map from $\mathbb{R} \to \mathbb{R}$ is not bijective? o.O

Could you maybe write down exactly what you mean mathematically, or alternatively point me to a resource where I can read about this? I don't think I fully understand exactly what issue you're pointing to here, but really feel like this is something I should know so very keen on learning more.

Reply:
Btw just implemented sampling from the following model:
``` julia
@model function demo_ordered(μ)
    k = length(μ)
    σ² ~ filldist(truncated(Normal(), lower=0), k)
    x ~ ordered(MvNormal(μ, Diagonal(σ²)))
    return (; σ², x)
end

k = 2
num_samples = 10_000

# Sample using NUTS.
μ = zeros(k)
σ² = ones(k)

model = fix(demo_ordered(μ); σ²)
chain = sample(model, NUTS(), num_samples)
xs_chain = permutedims(Array(chain))

# Rejection sampling.
d = MvNormal(μ, Diagonal(σ²))
xs_exact = mapreduce(hcat, 1:num_samples) do _
    xs = [rand(marginal(d, 1))]
    for i = 2:k
        x_prev = xs[end]
        while true
            x = rand(marginal(d, i))
            if x >= x_prev
                push!(xs, x)
                break
            end
        end
    end
    return xs
end

display(chain)

qts = [0.05, 0.25, 0.5, 0.75, 0.95]
qs_chain = mapslices(xs_chain; dims=2) do xs
    quantile(xs, qts)
end
qs_exact = mapslices(xs_exact; dims=2) do xs
    quantile(xs, qts)
end
```

and even that fails to produce consistent results :confused:

Are these two really supposed to be the same? 

Here's the output of the above:
```julia
┌ Info: Found initial step size
└   ϵ = 1.7000000000000002
Chains MCMC chain (10000×15×1 Array{Float64, 3}):

Iterations        = 1001:1:11000
Number of chains  = 1
Samples per chain = 10000
Wall duration     = 1.59 seconds
Compute duration  = 1.59 seconds
parameters        = x[1], x[2], x[3]
internals         = lp, n_steps, is_accept, acceptance_rate, log_density, hamiltonian_energy, hamiltonian_energy_error, max_hamiltonian_energy_error, tree_depth, numerical_error, step_size, nom_step_size

Summary Statistics
  parameters      mean       std      mcse     ess_bulk    ess_tail      rhat   ess_per_sec 
      Symbol   Float64   Float64   Float64      Float64     Float64   Float64       Float64 

        x[1]    0.0156    1.0073    0.0089   12830.2088   7634.0339    1.0003     8069.3137
        x[2]    1.6587    2.6114    0.0269   11976.8423   7526.6856    1.0000     7532.6052
        x[3]    3.2724    3.3218    0.0345   10898.1739   7911.5172    1.0001     6854.1974

Quantiles
  parameters      2.5%     25.0%     50.0%     75.0%     97.5% 
      Symbol   Float64   Float64   Float64   Float64   Float64 

        x[1]   -1.9646   -0.6662    0.0111    0.6789    1.9682
        x[2]   -1.2130    0.3388    1.2352    2.3719    7.3322
        x[3]   -0.3950    1.3836    2.6070    4.1849   11.4003

┌ Info: quantiles
│   qs_chain =
│    3×5 Matrix{Float64}:
│     -1.6457     -0.666194  0.0111229  0.678935  1.70693
│     -0.879387    0.338842  1.23522    2.37194   5.35761
│     -0.0182679   1.38357   2.60698    4.18492   8.67829
│   qs_exact =
│    3×5 Matrix{Float64}:
│     -1.6316    -0.683853  0.00858142  0.687207  1.64531
│     -0.506248   0.286727  0.888104    1.48044   2.36784
└      0.154898   0.905938  1.46605     2.03997   2.91713
```

As you can see, the quantiles are different.